File: processkeepalive/src/main/java/com/example/processkeepalive/service/RemoteService.java
package com.example.processkeepalive.service;

import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import androidx.annotation.IntDef;
import android.util.Log;
import android.widget.Toast;

import com.example.processkeepalive.aidl.IMyAidlInterface;

public class RemoteService extends Service {
    private static final String TAG = "RemoteService";
    private RemoteBinder myBinder;
    private ServiceConnection remoteServiceConnection;

    public RemoteService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return myBinder;
    }

    @Override
    public void onCreate() {
        Log.i(TAG, "onCreate: ");
        if (myBinder == null){
            myBinder = new RemoteBinder();
        }
        remoteServiceConnection = new RemoteServiceConnection();
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(TAG, "onStartCommand: ");
        bindService(new Intent(RemoteService.this, LocalService.class), remoteServiceConnection, Context.BIND_IMPORTANT);
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "onDestroy: ");
        super.onDestroy();
    }


    class RemoteBinder extends IMyAidlInterface.Stub{

        @Override
        public void basicTypes(int anInt, long aLong, boolean aBoolean, float aFloat, double aDouble, String aString) throws RemoteException {

        }

        @Override
        public String getName() throws RemoteException {
            return    "i am remote Service";
        }
    }

    class  RemoteServiceConnection implements ServiceConnection{

        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            Log.i(TAG, "onServiceConnected: 本地服务已经链接");
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            // 连接出现了异常断开了，RemoteService被杀掉了
            Toast.makeText(RemoteService.this, "本地服务被干掉", Toast.LENGTH_LONG).show();
            // 启动RemoteCastielService
            startService(new Intent(RemoteService.this, LocalService.class));
            bindService(new Intent(RemoteService.this, LocalService.class), remoteServiceConnection, Context.BIND_IMPORTANT);
        }
    }

}

File: myapplication/src/main/java/com/example/myapplication/service/RemoteService.kt
package com.example.myapplication.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import com.example.myapplication.IRemoteService

/**
@Author zhang tao
@Date   7/13/22 10:52 PM
@Desc
 */

private const val TAG = "RemoteService"

class RemoteService : Service() {

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, "onCreate: ")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "onStartCommand: ")
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onBind(intent: Intent?): IBinder? {
        Log.d(TAG, "onBind: ")
        return RemoteBinder()
    }

    inner class RemoteBinder : IRemoteService.Stub() {
        override fun basicTypes(anInt: Int, aLong: Long, aBoolean: Boolean, aFloat: Float, aDouble: Double, aString: String?) {
            Log.d(TAG, "basicTypes: ")
        }

        override fun getBasicInfo(): String {
            return "name name is lily, age == 18"
        }
    }
}